cross-chain/bob/contracts/OptimismMintableUpgradableTBTC.sol

```diff
@@ -0,0 +1,337 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity 0.8.15;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "./legacy/OptimismMintableUpgradableERC20.sol";
+
+/// @title OptimismMintableUpgradableTBTC
+/// @notice Canonical L2/sidechain token implementation. tBTC token is minted on
+///         L1 and locked there to be moved to L2/sidechain. By deploying
+///         a canonical token on each L2/sidechain, we can ensure the supply of
+///         tBTC remains sacrosanct, while enabling quick, interoperable
+///         cross-chain bridges and localizing ecosystem risk.
+///
+///         This contract is flexible enough to:
+///         - Delegate minting authority to a native bridge on the chain, if
+///           present.
+///         - Delegate minting authority to a short list of ecosystem bridges.
+///         - Have mints and burns paused by any one of n guardians, allowing
+///           avoidance of contagion in case of a chain- or bridge-specific
+///           incident.
+///         - Be governed and upgradeable.
+///
+///         The token is burnable by the token holder and supports EIP2612
+///         permits. Token holder can authorize a transfer of their token with
+///         a signature conforming EIP712 standard instead of an on-chain
+///         transaction from their address. Anyone can submit this signature on
+///         the user's behalf by calling the permit function, paying gas fees,
+///         and possibly performing other actions in the same transaction.
+///         The governance can recover ERC20 and ERC721 tokens sent mistakenly
+///         to OptimismMintableUpgradableTBTC token contract.
+/// @custom:oz-upgrades-from OptimismMintableUpgradableERC20
+contract OptimismMintableUpgradableTBTC is
```

Comment:
looks excellent and I am learning a lot, just wondering if do we need to include the `MisfundRecovery` functionality, `recoverERC20()` and 721, to recover funds or it was deprecated?:

MisfundRecovery.sol

```
import "@thesis/solidity-contracts/contracts/token/MisfundRecovery.sol";

contract TBTC is ERC20WithPermit, MisfundRecovery {
```

Reply:
got it